src/Data/Bin/Bin0.agda

```diff
@@ -0,0 +1,686 @@
+open import Relation.Nullary          using (yes; no)
+open import Relation.Nullary.Negation using (contradiction)
+
+open PE.≡-Reasoning
+import Algebra.Properties.Semigroup NatP.+-semigroup as Of+ℕ-semigroup
+
+
+
+--****************************************************************************
+data Bin : Set where
+               0#    : Bin
+               2suc  : Bin → Bin    -- \n→ 2*(1+n)  arbitrary nonzero even
+               suc2* : Bin → Bin    -- \n→ 1 + 2n   arbitrary odd
+--
+-- This representation is unique for each natural number.
+
+
+-- Decidable equality on Bin -------------------------------------------------
+
+2suc-injective : Injective _≡_ _≡_ 2suc
+2suc-injective {0#}      {0#}      _    =  refl
+2suc-injective {0#}      {2suc _}  ()
+2suc-injective {0#}      {suc2* _} ()
+2suc-injective {2suc _}  {0#}      ()
+2suc-injective {2suc _}  {2suc _}  refl =  refl
+2suc-injective {2suc _}  {suc2* _} ()
+2suc-injective {suc2* _} {0#}      ()
+2suc-injective {suc2* _} {2suc _}  ()
+2suc-injective {suc2* _} {suc2* _} refl =  refl
+
+suc2*-injective : Injective _≡_ _≡_ suc2*
+suc2*-injective {0#}      {0#}      _    =  refl
+suc2*-injective {0#}      {2suc _}  ()
+suc2*-injective {0#}      {suc2* _} ()
+suc2*-injective {2suc _}  {0#}      ()
+suc2*-injective {2suc _}  {2suc _}  refl =  refl
+suc2*-injective {2suc _}  {suc2* _} ()
+suc2*-injective {suc2* _} {0#}      ()
+suc2*-injective {suc2* _} {2suc _}  ()
+suc2*-injective {suc2* _} {suc2* _} refl =  refl
+
+
+_≟_ :  Decidable (_≡_ {A = Bin})
+
+0#        ≟ 0#        =  yes refl
+0#        ≟ (2suc _)  =  no λ()
+0#        ≟ (suc2* _) =  no λ()
+(2suc _)  ≟ 0#        =  no λ()
+(2suc x)  ≟ (2suc y)  with x ≟ y
+...                   | yes eq =  yes (cong 2suc eq)
+...                   | no ne  =  no (ne ∘ 2suc-injective)
+
+(2suc _)  ≟ (suc2* _) =  no λ()
+(suc2* _) ≟ 0#        =  no λ()
+(suc2* _) ≟ (2suc _)  =  no λ()
+(suc2* x) ≟ (suc2* y) with x ≟ y
+...                   | yes eq =  yes (cong suc2* eq)
+...                   | no ne  =  no (ne ∘ suc2*-injective)
+
+2suc≢0 :  {x : Bin} → 2suc x ≢ 0#
+2suc≢0 ()
+
+suc2*≢0 :  {x : Bin} → suc2* x ≢ 0#
+suc2*≢0 ()
+
+---------------------------------------------------------------
+size : Bin → ℕ       -- (number of constructors) - 1.
+                       -- It is used in some termination proofs
+size 0#        = 0
+size (2suc x)  = 1+ (size x)
+size (suc2* x) = 1+ (size x)
+
+|x|≡0⇒x≡0 :  {x : Bin} → size x ≡ 0 → x ≡ 0#
+|x|≡0⇒x≡0 {0#}      _ =  refl
+|x|≡0⇒x≡0 {2suc _}  ()
+|x|≡0⇒x≡0 {suc2* _} ()
+
+--============================================================================
+-- Arithmetic operations on Bin and their properties.
+
+suc : Bin → Bin
+suc 0#        =  suc2* 0#
+suc (2suc x)  =  suc2* (suc x)   -- 1 + 2(1+x)
+suc (suc2* x) =  2suc x          -- 1 + 1 + 2x =  2*(1+x)
+
+1# = suc 0#;   2# = suc 1#;   3# = suc 2#;   4# = suc 3#;   5# = suc 4#
```

Comment:
These would be better defined as patterns or even better via a Data.Bin.Literals module

Reply:
Do you mean to add the module  Data.Bin.Literals.agda  and to put there

1# = suc 0#
2# = suc 1#
...
?